File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode p1 = list1;
        ListNode p2 = list2;
        ListNode cur; 
        ListNode head;

        if (p1 != null && p2 != null) {
            if (p1.val < p2.val) {
                head = p1;
                p1 = p1.next;
            } else {
                head = p2;
                p2 = p2.next;
            }
        } else if (p1 == null && p2 != null) {
            return p2;
        } else if (p1 != null && p2 == null) {
            return p1;
        } else {
            return null;
        }

        cur = head;

        while (p1 != null & p2 != null) {
            if (p1.val < p2.val) {
                cur.next = p1;
                p1 = p1.next;
            } else {
                cur.next = p2;
                p2 = p2.next;
            }
            cur = cur.next;
        }
        
        if (p1 == null) {
            cur.next = p2;
        } else {
            cur.next = p1;
        }

        return head;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        count = 1
        node = head
        while node.next:
            count += 1
            node = node.next

        nodes = [None] * ((count - 1) // 2)
        # iterate through linked list again to get switch nodes 
        node = head 
        i = 1
        ptr = 0
        end = None
        while node.next:            
            if count - i <= (count - 1) // 2: 
                temp = node
                nodes[(len(nodes) - 1) - ptr] = node
                ptr += 1
                node = node.next
                if ptr == 1:
                    end = temp 
            else: 
                node = node.next
            i += 1
        
        # verify correct nodes were gotten
        lo = head
        for n in nodes:
            node = n.next
            print("trying to do " + str(lo.val) + " then " + str(node.val))
            temp = lo.next
            lo.next = node
            node.next = temp
            lo = temp
        
        if end:
            end.next = None